sycl/doc/design/DeviceAspectTraitDesign.md

```diff
@@ -0,0 +1,128 @@
+that in this section we refer to aspects as their integral representation as
+specified in the device headers rather than by the names specified in the
+[SYCL 2020 specification][1].
+
+For each target $t$ in `-fsycl-targets`, let $A^{any}_t$ be the set of aspects
+supported by any device supporting $t$ and let $A^{all}_t$ be the set of aspects
+supported by all devices supporting $t$. These sets are defined as follows:
+* If $t$ has an entry in the configuration file, $A^{all}_t$ is the same as the
+`aspects` list in that entry.
+* If $t$ does not have an entry in the configuration file, $A^{all}_t$ is empty.
+* If $t$ has an entry in the configuration file and the entry has a value
+`may_support_other_aspects` set to `false`, $A^{any}_t$ is the same as the
+`aspects` list in that entry.
+* If $t$ does not have an entry the configuration file or the entry has a value
+`may_support_other_aspects` set to `true`, $A^{any}_t$ is the set of all
+aspects.
+
+For example, the target `intel_gpu_dg1` is supported by a specific device (DG1)
+and as such would have an entry in the configuration file with `aspects` being
+the set of aspects that device supports. Likewise it would have
+`may_support_other_aspects` set to `false` as there will be no other devices
+supporting this target, meaning there will never be any devices supporting
+the target and supporting anything not in `aspects`.  In contrast, the target
+`nvidia_gpu_sm_80` is supported by CUDA devices with `sm_80` architecture or
+newer, so its entry in the configuration file would have
+`may_support_other_aspects` set to `true` to indicate that there could be future
+devices that support aspects not in `aspects`, while it is known that all
+current and future devices must support the aspects in `aspects`.  Lastly, the
+default JIT SPIR-V target (`spir64`) should not have an entry in the
+configuration file as it cannot guarantee anything about the devices supporting
+the target.
+
+When compiling a SYCL program, where $[t1, t2, \ldots, tn]$ are the $n$ targets
+specified in `-fsycl-targets` including any targets implicitly added by the
+driver, the driver defines the following macros:
+* `__SYCL_ALL_DEVICES_HAVE_`$i$`__` as `1` for all $i$ in
+${\bigcap}^n_{k=1} A^{all}_{tk}$.
+* `__SYCL_ANY_DEVICE_HAS_ANY_ASPECT__` as `1` if
+${\bigcup}^n_{k=1} A^{any}_{tk}$ is the set of all aspects.
+* `__SYCL_ANY_DEVICE_HAS_`$j$`__` as `1` for all $j$ in
+${\bigcup}^n_{k=1} A^{any}_{tk}$ if `__SYCL_ANY_DEVICE_HAS_ANY_ASPECT__` was not
+defined.
+
+Note that the need for the `__SYCL_ANY_DEVICE_HAS_ANY_ASPECT__` macro is
+due to the special case where the driver finds no configuration for a target and
+must assume that there exists some device that supports any given aspect. Since
+the driver has no way of knowing all possible aspects, we use a catch-all macro
+to denote this case instead. This is not needed for $A^{all}_t$ for any target
+$t$, as the driver will always know all relevant aspects.
```

Comment:
This is kind of a minor point, but the last sentence seems a little confusing to me.  It seems to me that the reason we do not need any special macro for the ` $A^{all}_t$` case is because ` $A^{all}_t$` is the empty set when there is no configuration entry for `t`.

Reply:
Re-reading it, I completely agree. I have clarified it. Do you think it is clearer now?